common/Geometry2d/Pose.hpp

```diff
@@ -0,0 +1,258 @@
+#pragma once
+
+#include "Point.hpp"
+#include "TransformMatrix.hpp"
+
+namespace Geometry2d {
+
+/**
+ * Represents a pose in 2d space: (x, y, theta). Uses double-precision floating
+ * point numbers for the coordinates, and can be used to represent the position
+ * and orientation of a robot in the plane.
+ */
+class Pose {
+public:
+    /**
+     * Default constructor - zero-initialize
+     */
+    Pose() { _pose = Eigen::Vector3d::Zero(); }
+
+    /**
+     * Point-heading constructor
+     */
+    Pose(Point position, double heading)
+        : _pose(position.x(), position.y(), heading) {}
+
+    /**
+     * Implicit conversion from Eigen::Vector3d
+     */
+    Pose(const Eigen::Vector3d& other) : _pose(other) {}
+
+    /**
+     * Copy-constructor - default
+     */
+    Pose(const Pose& other) = default;
+
+    /**
+     * Compute the pose specified using this pose as coordinates in a frame of
+     * reference specified by `other`, in the global space.
+     *
+     * i.e.
+     *
+     * x:
+     *       |  x
+     *       | /
+     *       |/
+     * ------|-------
+     *       |
+     *       |
+     *       |
+     *
+     * y:
+     *     |
+     *     |
+     * ----|----
+     *     |\
+     *     | y
+     *
+     * y.withOrigin(x):
+     *       |  x
+     *       | / \
+     *       |/   y
+     * ------|------
+     *       |
+     */
+    Pose withOrigin(Pose other) const {
+        double sh = std::sin(other.heading()), ch = std::cos(other.heading());
+        Point rotated = position().rotated(other.heading());
+        return other + Pose(rotated, heading());
+    }
+
+    /**
+     * Implicit conversion to Eigen::Vector3d
+     */
+    operator Eigen::Vector3d() const { return _pose; }
+
+    /**
+     * Calculate a TransformMatrix corresponding to using this pose as the
+     * origin of the coordinate system.
+     */
+    TransformMatrix transform() const {
+        return TransformMatrix(position(), heading());
+    }
+
+    /**
+     * Accessors
+     */
+    Point position() const { return Point(_pose.template block<2, 1>(0, 0)); }
+    double heading() const { return _pose(2); }
```

Comment:
same as comment above: make a `double&` version so that you can do `pose.heading() += M_PI` or something?

Reply:
or alternatively provide a setter?

Reply:
Read above comment chain. I don't want it to work for one accessor and not the other, and implementation for Point accessor is nontrivial.